doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,278 @@
+rather than sending to the bindings to the storage controller and waiting to
+hear that they have been durably persisted.
+
+The details are murky. There is a draft PR ([#11883](https://github.com/MaterializeInc/materialize/pull/11883))
+with more specifics.
+
+TODO: describe how to handle the Debezium transaction metadata topic.
+
+### Storage API changes
+
+Each source in the system is identified by a [`GlobalId`].  The global ID will
+refer to the reclocked collection in the following commands:
+
+  * [`StorageCommand::CreateSources`](https://dev.materialize.com/api/rust/mz_dataflow_types/client/enum.StorageCommand.html#variant.CreateSources)
+  * [`StorageCommand::AllowCompaction`](https://dev.materialize.com/api/rust/mz_dataflow_types/client/enum.StorageCommand.html#variant.AllowCompaction)
+  * [`StorageCommand::Insert`](https://dev.materialize.com/api/rust/mz_dataflow_types/client/enum.StorageCommand.html#variant.Insert)
+  * [`StorageComand::DurabilityFrontierUpdates`](https://dev.materialize.com/api/rust/mz_dataflow_types/client/enum.StorageCommand.html#variant.DurabilityFrontierUpdates)
+
+[`StorageCommand::RenderSources`](https://dev.materialize.com/api/rust/mz_dataflow_types/client/enum.StorageCommand.html#variant.RenderSources) will evolve to allow clients to specify *which* of the collections associated with the source to render:
+
+```rust
+/// Identifies a collection in the storage layer.
+struct StorageCollectionId {
+    /// The ID of the source with which the collection is associated.
+    source_id: GlobalId
+    /// Which collection associated with the source to choose.
+    collection_id: CollectionType,
+}
+
+/// Specifies the type of a collection associated with a source.
+enum CollectionType {
+    /// The collection containing the reclocked upstream data.
+    Reclocked,
+    /// The collection containing the reclocking mappings.
+    Remapping,
+}
+```
+
+### SQL changes
+
+The `CREATE SOURCE` statement will be extended with an `EXPOSE PROGRESS
+AS <name>` clause. For example:
+
+```sql
+CREATE SOURCE foo
+FROM KAFKA BROKER '...'
+EXPOSE PROGRESS AS my_foo_progress
+```
+
+Materialize will make the contents of the `remap` collection for the source
+visible under the specified name via a relation whose structure depends on the
+type of the source:
+
+* File sources
+
+  | mz_timestamp | offset |
+  |--------------|--------|
+  | ...          | ...    |
+
+* Kafka sources
+
+  | mz_timestamp | partition | offset |
+  |--------------|-----------|--------|
+  | ...          | ...       | ...    |
+
+* Kinesis sources
+
+  | mz_timestamp | shard_id | sequence_number |
+  |--------------|----------|-----------------|
+  | ...          | ...      | ...             |
+
+* PostgreSQL sources
+
+  | mz_timestamp | lsn |
+  |--------------|-----|
+  | ...          | ... |
+
+* S3 sources
+
+  | mz_timestamp | key | offset |
+  |--------------|-----|--------|
+  | ...          | ... | ...    |
```

Comment:
I'm not an expert, but I don't think offset makes sense for S3 in the way that it does for Kafka (as in I don't think S3 has a notion of offset explicitly). It does have byte-range fetches so perhaps we could leverage that, but unfortunately I haven't come up with something nicer to call it 😞

Reply:
Discussed in person with @benesch , we like `byte_offset`!

Reply:
Updated!